source/mqtt311_decoder.c

```diff
@@ -0,0 +1,206 @@
+        struct aws_byte_cursor whole_packet_cursor = *data;
+        whole_packet_cursor.len = fixed_header_length + packet_header.remaining_length;
+
+        /* advance the external, mutable data cursor to the start of the next packet */
+        aws_byte_cursor_advance(data, whole_packet_cursor.len);
+
+        /*
+         * if this fails, the decoder goes into an error state.  If it succeeds we'll loop again into the same state
+         * because we'll be back at the beginning of the next packet (if it exists).
+         */
+        enum aws_mqtt_packet_type packet_type = aws_mqtt_get_packet_type(whole_packet_cursor.ptr);
+        return aws_result_to_mqtt311_decoding_directive(
+            s_aws_mqtt311_decoder_safe_packet_handle(decoder, packet_type, whole_packet_cursor));
+    }
+
+    uint8_t byte = *data->ptr;
+    aws_byte_cursor_advance(data, 1);
+    aws_byte_buf_append_byte_dynamic(&decoder->packet_buffer, byte);
+
+    decoder->state = AWS_MDST_READ_REMAINING_LENGTH;
+
+    return AWS_MDD_CONTINUE;
+}
+
+static enum aws_mqtt311_decoding_directive s_handle_decoder_read_remaining_length(
+    struct aws_mqtt311_decoder *decoder,
+    struct aws_byte_cursor *data) {
+    AWS_FATAL_ASSERT(decoder->total_packet_length == 0);
+    if (data->len == 0) {
+        return AWS_MDD_OUT_OF_DATA;
+    }
+
+    uint8_t byte = *data->ptr;
+    aws_byte_cursor_advance(data, 1);
+    aws_byte_buf_append_byte_dynamic(&decoder->packet_buffer, byte);
+
+    struct aws_byte_cursor vli_cursor = aws_byte_cursor_from_buf(&decoder->packet_buffer);
+    aws_byte_cursor_advance(&vli_cursor, 1);
+
+    size_t remaining_length = 0;
+    int vli_result = aws_mqtt311_decode_remaining_length(&vli_cursor, &remaining_length);
+
+    /* anything other than a short buffer error (not enough data yet) is a terminal error */
+    if (vli_result == AWS_OP_ERR && aws_last_error() != AWS_ERROR_SHORT_BUFFER) {
+        return AWS_MDD_PROTOCOL_ERROR;
+    }
+
+    /*
+     * If we successfully decoded a variable-length integer, we now know exactly how many bytes we need to receive in
+     * order to have the full packet.
+     */
+    if (vli_result == AWS_OP_SUCCESS) {
+        decoder->total_packet_length = remaining_length + decoder->packet_buffer.len;
+        AWS_FATAL_ASSERT(decoder->total_packet_length > 0);
+        decoder->state = AWS_MDST_READ_BODY;
+    }
+
+    return AWS_MDD_CONTINUE;
+}
+
+static enum aws_mqtt311_decoding_directive s_handle_decoder_read_body(
+    struct aws_mqtt311_decoder *decoder,
+    struct aws_byte_cursor *data) {
+    AWS_FATAL_ASSERT(decoder->total_packet_length > 0);
+
+    size_t buffer_length = decoder->packet_buffer.len;
+    size_t amount_to_read = aws_min_size(decoder->total_packet_length - buffer_length, data->len);
+
+    struct aws_byte_cursor copy_cursor = aws_byte_cursor_advance(data, amount_to_read);
+    if (aws_byte_buf_append_dynamic(&decoder->packet_buffer, &copy_cursor)) {
+        return AWS_MDD_PROTOCOL_ERROR;
```

Comment:
trivial: you're checking this append_dynamic, but not others

I'd remove the error-checking, failed allocations are fatal